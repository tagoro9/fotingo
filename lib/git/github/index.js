'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _child_process = require('child_process');

var _child_process2 = _interopRequireDefault(_child_process);

var _fs = require('fs');

var _fs2 = _interopRequireDefault(_fs);

var _github = require('github');

var _github2 = _interopRequireDefault(_github);

var _ramda = require('ramda');

var _ramda2 = _interopRequireDefault(_ramda);

var _util = require('../util');

var _util2 = require('../../util');

var _error = require('../../error');

var _reporter = require('../../reporter');

var _reporter2 = _interopRequireDefault(_reporter);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

var github = new _github2['default']({});

var getCurrentUser = _ramda2['default'].compose(_ramda2['default'].partial(_ramda2['default'].__, [{}]), _util2.promisify)(github.users.get);

var createPullRequest = _ramda2['default'].composeP(_ramda2['default'].compose(_util2.wrapInPromise, _ramda2['default'].prop('data')), (0, _util2.promisify)(github.pullRequests.create));
var addLabels = _ramda2['default'].composeP(_ramda2['default'].compose(_util2.wrapInPromise, _ramda2['default'].prop('data')), (0, _util2.promisify)(github.issues.addLabels));
var getLabels = _ramda2['default'].composeP(_ramda2['default'].compose(_util2.wrapInPromise, _ramda2['default'].prop('data')), (0, _util2.promisify)(github.issues.getLabels));

var authenticate = _ramda2['default'].compose(_util2.wrapInPromise, function (token) {
  return github.authenticate(token);
}, _ramda2['default'].set(_ramda2['default'].lensProp('token'), _ramda2['default'].__, { type: 'token' }));

var readUserToken = _ramda2['default'].compose(_ramda2['default'].partial(_reporter2['default'].question, [{ question: 'Introduce a Github personal token' }]), (0, _util2.debugCurried)('github', 'Github account not set'));

var authenticateAndGetCurrentUser = _ramda2['default'].composeP(_ramda2['default'].compose((0, _error.catchPromiseAndThrow)('github', function (e) {
  (0, _util2.debug)('github', 'Error when authenticating: ' + e.message);
  switch (e.code) {
    case '500':
      return _error.errors.github.cantConnect;
    default:
      return _error.errors.github.tokenInvalid;
  }
}), getCurrentUser), authenticate);

// Object -> Object -> String
var buildPullRequestTitle = _ramda2['default'].ifElse(_ramda2['default'].isNil, _ramda2['default'].compose(_ramda2['default'].concat(_ramda2['default'].__, '\n'), _ramda2['default'].prop('name'), _ramda2['default'].nthArg(1)), function (issue) {
  return _util.ISSUE_TYPES[issue.fields.issuetype.name] + '/' + issue.key + ' ' + _ramda2['default'].take(60, issue.fields.summary) + '\n';
});

// Array -> String
var buildPullRequestBody = _ramda2['default'].compose(_ramda2['default'].join('\n'), _ramda2['default'].map(_ramda2['default'].converge(_ramda2['default'].concat, [_ramda2['default'].compose(_ramda2['default'].concat('* '), _ramda2['default'].prop('header')), _ramda2['default'].compose(_ramda2['default'].ifElse(_ramda2['default'].isNil, _ramda2['default'].always(''), _ramda2['default'].replace(/\n|^/g, '\n  ')), _ramda2['default'].prop('body'))])));
// Array -> String
var buildPullRequestFooter = function () {
  function buildPullRequestFooter(issueRoot) {
    return _ramda2['default'].compose(_ramda2['default'].ifElse(_ramda2['default'].isEmpty, _ramda2['default'].always(''), _ramda2['default'].compose(_ramda2['default'].concat('\nFixes '), _ramda2['default'].join(', '))), _ramda2['default'].map(function (_ref) {
      var raw = _ref.raw,
          issue = _ref.issue;
      return '[' + raw + '](' + issueRoot + issue + ')';
    }), _ramda2['default'].uniqBy(_ramda2['default'].prop('issue')));
  }

  return buildPullRequestFooter;
}();

// Object -> Object -> String
var buildPullRequestDescription = function () {
  function buildPullRequestDescription(issueRoot) {
    return _ramda2['default'].converge(_ramda2['default'].compose(_util2.wrapInPromise, _ramda2['default'].unapply(_ramda2['default'].join('\n'))), [buildPullRequestTitle, _ramda2['default'].compose(buildPullRequestBody, _ramda2['default'].prop('commits'), _ramda2['default'].nthArg(1)), _ramda2['default'].compose(buildPullRequestFooter(issueRoot), _ramda2['default'].prop('issues'), _ramda2['default'].nthArg(1))]);
  }

  return buildPullRequestDescription;
}();

// String -> String -> *
var writeFile = _ramda2['default'].curryN(2, function (file, content) {
  return (0, _util2.promisify)(_fs2['default'].writeFile)(file, content, 'utf8');
});
// String -> * -> String
var readFile = function () {
  function readFile(file) {
    return function () {
      return (0, _util2.promisify)(_fs2['default'].readFile)(file, 'utf8');
    };
  }

  return readFile;
}();
// String -> String -> String
var deleteFile = function () {
  function deleteFile(file) {
    return function (content) {
      return (0, _util2.promisify)(_fs2['default'].unlink)(file).then(function () {
        return (0, _util2.wrapInPromise)(content);
      });
    };
  }

  return deleteFile;
}();
// * -> String
var editFile = function () {
  function editFile(file) {
    return function () {
      return new Promise(function (resolve, reject) {
        var vim = _child_process2['default'].spawn('vim', [file], { stdio: 'inherit' });
        vim.on('exit', _ramda2['default'].ifElse(_ramda2['default'].equals(0), resolve, reject));
      });
    };
  }

  return editFile;
}();

// Object -> String -> String
var allowUserToEditPullRequest = function () {
  function allowUserToEditPullRequest(description) {
    var prFile = '/tmp/fotingo-pr-' + Date.now();
    return _ramda2['default'].composeP(_ramda2['default'].compose(_util2.wrapInPromise, _ramda2['default'].trim), deleteFile(prFile), readFile(prFile), editFile(prFile), writeFile(prFile))(description);
  }

  return allowUserToEditPullRequest;
}();

var submitPullRequest = _ramda2['default'].curryN(4, function (config, project, branchInfo, description) {
  return createPullRequest({
    owner: config.get(['github', 'owner']),
    repo: project,
    head: branchInfo.name,
    base: config.get(['github', 'base']),
    title: _ramda2['default'].compose(_ramda2['default'].head, _ramda2['default'].split('\n'))(description),
    body: _ramda2['default'].compose(_ramda2['default'].join('\n'), _ramda2['default'].tail, _ramda2['default'].split('\n'))(description)
  });
});

var addLabelsToPullRequest = _ramda2['default'].curryN(4, function (config, project, branchInfo, pullRequest) {
  return _ramda2['default'].composeP(_ramda2['default'].compose(_util2.wrapInPromise, _ramda2['default'].always(pullRequest)), addLabels)({
    owner: config.get(['github', 'owner']),
    repo: project,
    number: pullRequest.number,
    labels: branchInfo.labels
  });
});

exports['default'] = {
  init: function () {
    function init(config) {
      return function () {
        (0, _util2.debug)('github', 'Initializing Github api');

        var doLogin = _ramda2['default'].composeP(authenticateAndGetCurrentUser, config.update(['github', 'token']), readUserToken);

        var configPromise = _ramda2['default'].isNil(config.get(['github', 'owner'])) ? _ramda2['default'].composeP(config.update(['github', 'owner']), _reporter2['default'].question)({ question: 'What\'s the github repository owner?' }) : (0, _util2.wrapInPromise)(config.get(['github', 'owner']));

        return configPromise.then(function () {
          if (config.isGithubLoggedIn()) {
            (0, _util2.debug)('github', 'User token is present. Using current authentication');
            return authenticateAndGetCurrentUser(config.get(['github', 'token']))
            // TODO differentiate error codes so only login is attempted when tokenInvalid
            ['catch'](_ramda2['default'].composeP(doLogin, (0, _util2.debugCurriedP)('github', 'Current authentication failed. Attempting login')));
          }
          (0, _util2.debug)('github', 'No user token present. Attempting login');
          return doLogin();
        });
      };
    }

    return init;
  }(),
  // Object -> Array -> Promise
  createPullRequest: _ramda2['default'].curryN(5, function (config, project, issue, issueRoot, branchInfo) {
    return _ramda2['default'].composeP(_ramda2['default'].ifElse(_ramda2['default'].isEmpty, (0, _error.throwControlledError)(_error.errors.github.pullRequestDescriptionInvalid),
    // Assign the PR link to all the issues that were created
    _ramda2['default'].composeP(_ramda2['default'].compose(_util2.wrapInPromise, _ramda2['default'].set(_ramda2['default'].lensProp('pullRequest'), _ramda2['default'].__, { branchInfo: branchInfo })), addLabelsToPullRequest(config, project, branchInfo), submitPullRequest(config, project, branchInfo))), (0, _util2.debugCurriedP)('github', 'Submitting pull request to github'), allowUserToEditPullRequest, (0, _util2.debugCurriedP)('github', 'Editing pull request description'), buildPullRequestDescription(issueRoot))(issue, branchInfo);
  }),
  checkAndGetLabels: _ramda2['default'].curryN(4, function (config, project, labels, branchInfo) {
    return _ramda2['default'].composeP(_ramda2['default'].compose(_ramda2['default'].set(_ramda2['default'].lensProp('labels'), _ramda2['default'].__, branchInfo), _ramda2['default'].map(_ramda2['default'].prop('name')), _ramda2['default'].filter(_ramda2['default'].compose(_ramda2['default'].contains(_ramda2['default'].__, labels), _ramda2['default'].prop('name')))), _ramda2['default'].unless(_ramda2['default'].compose(function (names) {
      return _ramda2['default'].all(_ramda2['default'].contains(_ramda2['default'].__, names), labels);
    }, _ramda2['default'].map(_ramda2['default'].prop('name'))), (0, _error.throwControlledError)(_error.errors.github.invalidLabel)), getLabels)({ owner: config.get(['github', 'owner']), repo: project });
  })
};