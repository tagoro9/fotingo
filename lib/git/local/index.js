'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _slicedToArray = function () { function sliceIterator(arr, i) { var _arr = []; var _n = true; var _d = false; var _e = undefined; try { for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"]) _i["return"](); } finally { if (_d) throw _e; } } return _arr; } return function (arr, i) { if (Array.isArray(arr)) { return arr; } else if (Symbol.iterator in Object(arr)) { return sliceIterator(arr, i); } else { throw new TypeError("Invalid attempt to destructure non-iterable instance"); } }; }();

var _nodegit = require('nodegit');

var _nodegit2 = _interopRequireDefault(_nodegit);

var _ramda = require('ramda');

var _ramda2 = _interopRequireDefault(_ramda);

var _conventionalCommitsParser = require('conventional-commits-parser');

var _conventionalCommitsParser2 = _interopRequireDefault(_conventionalCommitsParser);

var _error = require('../../error');

var _util = require('../util');

var _util2 = require('../../util');

var _package = require('../../../package.json');

var _package2 = _interopRequireDefault(_package);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

var fetchOptions = {
  callbacks: {
    certificateCheck: _ramda2['default'].always(1),
    credentials: _ramda2['default'].compose(_nodegit2['default'].Cred.sshKeyFromAgent, (0, _util2.debugCurried)('git', 'Getting authentication from SSH agent'),
    // TODO Detect ssh key not present
    _ramda2['default'].nthArg(1))
  }
};

var repository = null;

var getCurrentBranchName = _ramda2['default'].composeP(function (ref) {
  return _nodegit2['default'].Branch.name(ref);
}, function () {
  return repository.head();
});

// Commit -> Object
var transformCommit = _ramda2['default'].compose(_conventionalCommitsParser2['default'].sync, _ramda2['default'].invoker(0, 'message'));

// Object -> Array -> Array
var getIssues = _ramda2['default'].converge(_ramda2['default'].concat, [_ramda2['default'].compose(_ramda2['default'].ifElse(_ramda2['default'].isNil, _ramda2['default'].always([]), function (_ref) {
  var key = _ref.key;
  return [{ raw: '#' + key, issue: key }];
}), _ramda2['default'].nthArg(0)), _ramda2['default'].compose(_ramda2['default'].flatten, _ramda2['default'].map(_ramda2['default'].compose(_ramda2['default'].map(_ramda2['default'].pick(['raw', 'issue'])), _ramda2['default'].prop('references'))), _ramda2['default'].nthArg(1))]);

exports['default'] = {
  init: function () {
    function init(config, pathToRepo) {
      return function () {
        (0, _util2.debug)('git', 'Initializing ' + pathToRepo + ' repository');
        return _nodegit2['default'].Repository.open(pathToRepo).then(function (repo) {
          repository = repo;
          return Promise.resolve(undefined);
        })['catch']((0, _error.throwControlledError)(_error.errors.git.couldNotInitializeRepo, { pathToRepo: pathToRepo }));
      };
    }

    return init;
  }(),
  createBranchName: _util.createBranchName,
  createIssueBranch: _ramda2['default'].curryN(2, function (config, name) {
    (0, _util2.debug)('git', 'Creating branch for issue');

    var _config$get = config.get(['git']),
        remote = _config$get.remote,
        branch = _config$get.branch;

    (0, _util2.debug)('git', 'Fetching data from remote');
    // We should fetch -> co master -> reset to origin/master -> create branch
    return repository.fetch(remote, fetchOptions).then((0, _util2.debugCurriedP)('git', 'Getting local repository status')).then(function () {
      return repository.getStatus();
    }).then(_ramda2['default'].ifElse(_ramda2['default'].isEmpty, _ramda2['default'].identity, function () {
      return _nodegit2['default'].Stash.save((0, _util2.debugCurried)('git', 'Stashing changes', repository), repository.defaultSignature(), 'auto generated stash by ' + _package2['default'].name, _nodegit2['default'].Stash.FLAGS.INCLUDE_UNTRACKED);
    })).then(function () {
      return repository.getBranchCommit(remote + '/' + branch);
    }).then((0, _util2.debugCurriedP)('git', 'Creating new branch')).then(_ramda2['default'].compose((0, _error.catchPromiseAndThrow)('git', _error.errors.git.branchAlreadyExists), function (commit) {
      return repository.createBranch(name, commit);
    })).then(function () {
      return repository.checkoutBranch(name);
    });
  }),
  getCurrentBranchName: getCurrentBranchName,
  pushBranchToGithub: _ramda2['default'].converge(_ramda2['default'].composeP(function (_ref2) {
    var _ref3 = _slicedToArray(_ref2, 3),
        remote = _ref3[0],
        branch = _ref3[1],
        _ref3$ = _ref3[2],
        branchName = _ref3$.branchName,
        ref = _ref3$.ref;

    return remote.push([ref], fetchOptions).then(function () {
      return _nodegit2['default'].Branch.setUpstream(branch, remote.name() + '/' + branchName);
    });
  }, function () {
    for (var _len = arguments.length, promises = Array(_len), _key = 0; _key < _len; _key++) {
      promises[_key] = arguments[_key];
    }

    return Promise.all(promises);
  }), [_ramda2['default'].compose(function (remote) {
    return _nodegit2['default'].Remote.lookup(repository, remote);
  }, _ramda2['default'].prop('remote'), _ramda2['default'].invoker(1, 'get')(['git'])), function () {
    return repository.head();
  }, function () {
    return getCurrentBranchName().then(function (name) {
      return { branchName: name, ref: 'refs/heads/' + name + ':refs/heads/' + name };
    });
  }]),
  extractIssueFromCurrentBranch: function () {
    function extractIssueFromCurrentBranch() {
      return _ramda2['default'].composeP((0, _util2.debugCurriedP)('git', 'Extracting issue from current branch'), _ramda2['default'].compose(_util2.wrapInPromise, _util.getIssueIdFromBranch), getCurrentBranchName)();
    }

    return extractIssueFromCurrentBranch;
  }(),
  getBranchInfo: function () {
    function getBranchInfo(issue) {
      (0, _util2.debug)('git', 'Getting branch commit history');
      return Promise.all([repository.getHeadCommit(), repository.getBranchCommit('origin/master')]).then(_ramda2['default'].when(_ramda2['default'].compose(_ramda2['default'].not, _ramda2['default'].allUniq, _ramda2['default'].map(_ramda2['default'].compose(_ramda2['default'].toString, _ramda2['default'].invoker(0, 'id')))), (0, _error.throwControlledError)(_error.errors.git.noChanges))).then(function (_ref4) {
        var _ref5 = _slicedToArray(_ref4, 2),
            latestCommit = _ref5[0],
            latestMasterCommit = _ref5[1];

        return Promise.all([_nodegit2['default'].Merge.base(repository, latestCommit, latestMasterCommit).then(function (latestCommonCommit) {
          (0, _util2.debug)('git', 'Created history walker. Latest common commit: ' + latestCommonCommit);
          var historyWalker = repository.createRevWalk();
          var commitStopper = function () {
            function commitStopper(commit) {
              return !latestCommonCommit.equal(commit.id());
            }

            return commitStopper;
          }();
          historyWalker.push(latestCommit);
          return historyWalker.getCommitsUntil(commitStopper);
        }).then(function (commits) {
          return _ramda2['default'].compose(_ramda2['default'].reverse, _ramda2['default'].map(transformCommit), _ramda2['default'].init)(commits);
        }), getCurrentBranchName()]).then(function (_ref6) {
          var _ref7 = _slicedToArray(_ref6, 2),
              commits = _ref7[0],
              name = _ref7[1];

          return {
            name: name,
            commits: commits,
            issues: getIssues(issue, commits)
          };
        });
      });
    }

    return getBranchInfo;
  }()
};