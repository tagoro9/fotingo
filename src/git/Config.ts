/**
 * Git config
 */
export interface GitConfig {
  /**
   * Name of the remote to connect to
   */
  remote: string;
  /**
   * Name of the base branch to use when creating a new branch
   */
  baseBranch: string;

  /**
   * Template to use for creating branch names
   */
  branchTemplate: string;
}

export interface GithubConfig {
  authToken: string;
  baseBranch: string;
  owner: string;
  repo: string;
  pullRequestTemplate: string;
}
