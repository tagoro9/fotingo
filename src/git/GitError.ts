export enum GitErrorType {
  BRANCH_ALREADY_EXISTS,
  NOT_A_GIT_REPO,
}

export interface GitError {
  readonly code: GitErrorType;
  readonly message: string;
}

export class GitErrorImpl extends Error implements GitError {
  public readonly code: GitErrorType;

  constructor(message: string, code: GitErrorType) {
    super(message);
    Object.setPrototypeOf(this, new.target.prototype);
    this.code = code;
  }
}
