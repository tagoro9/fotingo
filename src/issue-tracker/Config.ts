export interface JiraConfig {
  user: {
    login: string;
    token: string;
  };
  releaseTemplate: string;
  root: string;
}
