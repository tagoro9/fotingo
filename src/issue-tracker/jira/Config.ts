export interface JiraConfig {
  root: string;
  user: {
    login: string;
    token: string;
  };
}
