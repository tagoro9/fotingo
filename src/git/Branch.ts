/**
 * Git branch handler
 */
import * as escapeStringRegexp from 'escape-string-regexp';
import {
  __,
  compose,
  converge,
  curryN,
  identity,
  ifElse,
  isNil,
  lt,
  mapObjIndexed,
  match,
  nth,
  nthArg,
  prop,
  reduce,
  replace,
  subtract,
  toPairs,
  toUpper,
} from 'ramda';
import { Issue, IssueType } from 'src/types';
import { GitConfig } from './Config';

const ISSUE_TYPE_TO_BRANCH_PREFIX: { [S in IssueType]: string } = {
  [IssueType.BUG]: 'b',
  [IssueType.FEATURE]: 'f',
  [IssueType.STORY]: 'f',
  [IssueType.SUB_TASK]: 'f',
  [IssueType.TASK]: 'c',
};

enum TemplateKey {
  ISSUE_SHORT_NAME = 'issue.shortName',
  ISSUE_KEY = 'issue.key',
  ISSUE_SANITIZED_SUMMARY = 'issue.sanitizedSummary',
}

const TEMPLATE_KEYS_TO_MATCHERS: { [S in TemplateKey]: string } = {
  [TemplateKey.ISSUE_SHORT_NAME]: '(\\w+)',
  [TemplateKey.ISSUE_KEY]: '(\\w+-\\d+)',
  [TemplateKey.ISSUE_SANITIZED_SUMMARY]: '([\\w_-]+)',
};

interface TemplateData {
  data: {
    [TemplateKey.ISSUE_SHORT_NAME]: string;
    [TemplateKey.ISSUE_KEY]: string;
    [TemplateKey.ISSUE_SANITIZED_SUMMARY]: string;
  };
  template: string;
}

/**
 * Given a git configuration and an issue build the branch template data
 * @param config Git configuration
 * @param issue Issue
 */
function getTemplateData(config: GitConfig, issue: Issue): TemplateData {
  return {
    data: {
      [TemplateKey.ISSUE_SHORT_NAME]: ISSUE_TYPE_TO_BRANCH_PREFIX[issue.type],
      [TemplateKey.ISSUE_KEY]: issue.key.toLowerCase(),
      [TemplateKey.ISSUE_SANITIZED_SUMMARY]: issue.sanitizedSummary,
    },
    template: config.branchTemplate,
  };
}

/**
 * Get a branch name for the passed configuration and issue
 * @param config Git configuration
 * @param issue Issue
 */
export const getName = curryN(
  2,
  compose(
    converge(
      reduce((msg: string, [k, v]: string[]) => replace(`{${k}}`, v, msg)),
      [prop('template'), compose(toPairs, prop('data'))],
    ),
    getTemplateData,
  ),
);

/**
 * Build a branch template regex from the current git config branch template
 * @param config Git configuration
 */
const buildBranchTemplateRegex = compose(
  reduce(
    (msg: string, [k, v]: [TemplateKey, string]) => replace(`{${k}}`, `${v}`, msg),
    __ as any,
    toPairs(TEMPLATE_KEYS_TO_MATCHERS),
  ),
  prop('branchTemplate'),
);

/**
 * Given a regex that matches a branchTemplate, build a map between the template keys and the
 * matching group index in the regex
 * @param regex Regex to match a branch name
 */
const getTemplateKeysMatchIndexMap = (regex: string): { [S in TemplateKey]: number } => {
  const indexMap = Object.values(TemplateKey).reduce(
    (acc, val: TemplateKey) => ({
      ...acc,
      [val]: (
        regex.match(escapeStringRegexp(TEMPLATE_KEYS_TO_MATCHERS[val])) || {
          index: -1,
        }
      ).index,
    }),
    {},
  );
  const orderedValues = Object.values(indexMap)
    .sort(subtract)
    .filter(lt(-1));
  return mapObjIndexed(i => 1 + orderedValues.indexOf(i), indexMap) as {
    [S in TemplateKey]: number;
  };
};

/**
 * Extract the issue id from the branch name given a git configuration
 * @param config Git configuration
 * @param branchName Branch name
 */
export const getIssueId = curryN(
  2,
  compose(
    converge(
      compose(
        ifElse(isNil, identity, toUpper),
        converge(nth, [compose(prop(TemplateKey.ISSUE_KEY), getTemplateKeysMatchIndexMap), match]),
      ),
      [buildBranchTemplateRegex, nthArg(1)],
    ),
  ),
);
