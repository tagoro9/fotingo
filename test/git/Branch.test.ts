/**
 * Branch tests
 */
import 'jest';
import { getIssueId, getName } from 'src/git/Branch';
import { GitConfig } from 'src/git/Config';
import data from '../lib/data';

const gitConfig: GitConfig = {
  baseBranch: 'master',
  branchTemplate: '{issue.key}.{issue.shortName}.{issue.sanitizedSummary}',
  remote: 'origin',
};

const issue = data.createIssue();

describe('Branch', () => {
  describe('getName', () => {
    test('replaces the placeholders with the issue data', () => {
      expect(getName(gitConfig, issue)).toMatchSnapshot();
    });

    test('ignores unknown template data', () => {
      expect(
        getName(
          {
            ...gitConfig,
            branchTemplate: '{does_not_exist}',
          },
          issue,
        ),
      ).toMatchSnapshot();
    });
  });

  describe('getIssueId', () => {
    test('extract the issue id from the branch name using the template config', () => {
      expect(getIssueId(gitConfig, 'TEST-1234.b.my_issue')).toBe('TEST-1234');
      expect(
        getIssueId(
          {
            ...gitConfig,
            branchTemplate: '{issue.shortName}/{issue.key}_{issue.sanitizedSummary}',
          },
          'f/TEST-1234_my_project',
        ),
      ).toBe('TEST-1234');
      expect(
        getIssueId(
          {
            ...gitConfig,
            branchTemplate: '{issue.key}',
          },
          'TEST-1234',
        ),
      ).toBe('TEST-1234');
      expect(getIssueId(gitConfig, 'doesnotmatchtheregex')).toBe(undefined);
    });
  });
});
