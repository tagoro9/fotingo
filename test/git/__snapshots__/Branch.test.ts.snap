// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Branch getName ignores unknown template data 1`] = `"{does_not_exist}"`;

exports[`Branch getName replaces the placeholders with the issue data 1`] = `"fotingo-3178.f.National"`;
